src/plugins/rule/commands.js

```diff
@@ -0,0 +1,196 @@
+            }
+        }
+
+        // Previous two loops might have changed |parent|.
+        parent = parent.object === 'document'
+            ? editor.value.document
+            : editor.value.document.getDescendant(parent.key)
+
+        // When using `editor.wrapBlock` slate will place blocks according to
+        // its own rules, which don't take into account legal parent-child
+        // relations, which may cause rule to be created as a child of e.g.
+        // a list, and then promptly removed by normalizations. To avoid this we
+        // only use `editor.wrapBlock` when we're sure output will be OK.
+        if (RULE_PARENT.includes(parent.type || parent.object)) {
+            editor.wrapBlock('rule')
+            editor.wrapBlock('rule_statement')
+            return
+        }
+
+        // At this point the selection begins and ends on direct children
+        // of |parent|, which means we can start unwrapping nodes out if it.
+
+        while (!RULE_PARENT.includes(parent.type || parent.object)) {
+            if (first === last) {
+                // Simple case, Slate can handle it on its own.
+                editor.unwrapNodeByKey(first.key)
+                parent = editor.value.document.getParent(first.key)
+            } else {
+                const firstInx = parent.nodes.indexOf(first)
+                const lastInx = parent.nodes.indexOf(last)
+
+                if (lastInx + 1 < parent.nodes.size) {
+                    editor.splitNodeByKey(parent.key, lastInx + 1)
+                }
+                if (firstInx > 0) {
+                    editor.splitNodeByKey(parent.key, firstInx)
+                }
+
+                first = last = editor.value.document.getParent(first.key)
+                parent = editor.value.document.getParent(first.key)
+            }
+        }
+
+        // At this point the selected range is directly under a legal parent
+        // of rule.
+
+        const start = parent.nodes.indexOf(first)
+        const end = parent.nodes.indexOf(last) + 1
+        const nodes = parent.nodes.slice(start, end)
+
+        // Note that nodes in |nodes| are still children of |parent|, so we
+        // can't pass them as to `Block.create`, we'll have to move them.
+        const problem = Block.create({ type: 'rule_statement' })
+        const rule = Block.create({ type: 'rule', nodes: [problem] })
+        editor.insertNodeByKey(parent.key, start, rule)
+
+        // Now we can move |nodes| into |problem|.
+        nodes.forEach((node, inx) => {
+            editor.moveNodeByKey(node.key, problem.key, inx)
+        })
+    })
+}
+
+/**
+ * Create a new child element, insert it into current rule, and collapse
+ * selection into it.
+ *
+ * Fails if no rule is selected.
+ *
+ * @param {Slate~Editor} editor
+ * @param {string} type         type of element to insert
+ * @param {string} ...previous  type of elements before the one to insert
+ */
+function insertInRule(editor, type, ...previous) {
```

Comment:
It would be easier to read if previous would be just an array of string since it will be converted to an array anyway

Reply:
While I agree that this is somewhat strange, there's a good reason for this. `type` and `...previous` when considered together in this order name (in reverse) elements which must appear before the element to be inserted (`type` is also considered a part of `previous` because the newly added node will be after all previous nodes of that type). This could be represented as a single array (i.e. `insertInRule(editor: Editor, previous: string[])`, with TS types added for clarity), but that doesn't signify that the array must be non-empty. To put it differently, this code will generate a compile time warning:

```js
function insertInRule(editor, type, ...previous) { ... }

insertInRule(editor)
```

while this code won't:

```js
function insertInRule(editor, previous_and_type) { ... }

insertInRule(editor, [])
```

As for why it's a spread array rather than a normal array, I think `insertInRule(editor, 'a', 'b', 'c')` conveys the intent better than `insertInRule(editor, 'a', ['b', 'c'])`, when `'a'`, `'b'`, and `'c'` should in fact be considered a single array.